### ISSUE TITLE
feval in lgb.cv giving unexpected results

### ISSUE THREAD

**Issue by goswami-rahul** [Microsoft/LightGBM#1483]
For bugs and unexpected issues, please provide following information, so that we could reproduce on our system.

## Environment info
Operating System: Ubuntu 16.04 LTS
CPU: Using Google Colab GPU runtime
C++/Python/R version: Python 3

## Error Message:
No error
## Reproducible examples
I used `lgb.cv` as: 
```python
cvresult = lgb.cv(alg.get_params(), lgbtrain, num_boost_round=3000, nfold=5,
                 early_stopping_rounds=80, verbose_eval=True, seed=42, metrics="multi_logloss",
                 feval=f1_scorer)
```
I have defined my `f1_scorer` (passed as `feval` to `lgv.cv`) function as:
```python
def f1_scorer(y_pred, y):
    y = y.get_label().astype("int")
    y_pred = y_pred.reshape((-1, 5)).argmax(axis=1) 
    return "F1_scorer", metrics.f1_score(y, y_pred, average="weighted"), True
```
I reshaped and argmaxed `y_pred` because I guess `y_pred` were probabilties predicted on cv.  
I have 5 classes in my multi-class model, and `y_pred`'s shape was 5 times of `y` (true labels).  
For e.g., if I have 10 examples, then   
`y.shape` is `(10,)`. I expected `y_pred.shape` to be `(10, 5)`, but it was `(50,)`. (I know it because `metrics.f1_score` threw error displaying the mismatched shapes of these).  
So, I thought reshaping it to `(-1, 5)` would do the trick.  
But it seems there was some mistake by me defining `f1_scorer`, because the `f1_score` doesn't increase, while there is a substantial decrease in `multi_logloss`. Here are my outputs:
```
[1]	cv_agg's multi_logloss: 1.48182 + 0.000107528	cv_agg's F1_scorer: 0.208046 + 0.00171022
[2]	cv_agg's multi_logloss: 1.37942 + 0.000209271	cv_agg's F1_scorer: 0.20873 + 0.0017343
[3]	cv_agg's multi_logloss: 1.30399 + 0.000401368	cv_agg's F1_scorer: 0.209169 + 0.00158037
[4]	cv_agg's multi_logloss: 1.23172 + 0.00047433	cv_agg's F1_scorer: 0.209576 + 0.00178056
[5]	cv_agg's multi_logloss: 1.16928 + 0.000577606	cv_agg's F1_scorer: 0.209329 + 0.00187392
[6]	cv_agg's multi_logloss: 1.11477 + 0.000623601	cv_agg's F1_scorer: 0.209316 + 0.001725
[7]	cv_agg's multi_logloss: 1.06698 + 0.000639912	cv_agg's F1_scorer: 0.209314 + 0.00166868
[8]	cv_agg's multi_logloss: 1.0246 + 0.000678841	cv_agg's F1_scorer: 0.209319 + 0.0018861
.
. # skipped some outputs
.
[150]	cv_agg's multi_logloss: 0.615405 + 0.00142301	cv_agg's F1_scorer: 0.209562 + 0.00165159
[151]	cv_agg's multi_logloss: 0.615341 + 0.00142724	cv_agg's F1_scorer: 0.209498 + 0.0015317
[152]	cv_agg's multi_logloss: 0.615274 + 0.0014286	cv_agg's F1_scorer: 0.209505 + 0.00161461
[153]	cv_agg's multi_logloss: 0.615205 + 0.00143131	cv_agg's F1_scorer: 0.209524 + 0.0016036
[154]	cv_agg's multi_logloss: 0.61514 + 0.00143731	cv_agg's F1_scorer: 0.20951 + 0.00160288
[155]	cv_agg's multi_logloss: 0.615072 + 0.00143254	cv_agg's F1_scorer: 0.209491 + 0.00158067
```
See, the `multi_logloss` came down from `1.5` to `0.6`, but the `f1_score` is constant.  
`cv` was eventually stopped due `early_stopping` on `f1_scorer`.  
What am I doing wrong?  (I suspect it's from the part where I reshaped `y_pred` in `f1_scorer`, but not sure)

**Comment by Laurae2** [comment]
@goswami-rahul A decrease in logloss doesn't guarantee F1 score increase.

Try to compute manually the F1 score from the matrix of TP / FP / TN / FN of the 1st iteration and the last iteration to check, you will notice it is an expected behavior of the F1 score.

In any case, for quick checking: can you provide the confusion matrix of the 1st and last iterations, along with their computed logloss and F1 scores?

**Comment by goswami-rahul** [comment]
Thanks for the quick response.
@Laurae2 How shall I provide confusion matrices of the first and last iterations of `lgb.cv`. I need to train them again using `model.fit` once with `n_estimators=1` and then `n_estimators=155`, right?

**Comment by Laurae2** [comment]
@goswami-rahul Train a single model using lgb.train and provide out of fold confusion matrix.

**Comment by goswami-rahul** [comment]
Can you please clarify what exactly I have to do to produce confusion matrices from lgb.train? A small example code snippet would help :)  
Another info - I submitted the predictions from this model to the competition, where the evaluation was `f1 score with weighted average`, and it shows score `0.695` on Leaderboard (higher is better). So, I guess this indicates that f1 score is indeed increasing?

**Comment by goswami-rahul** [comment]
Can you also please confirm if my `f1_scorer` function is defined correct? I mean, I did it right to reshape it to `(-1, 5)` right? It was totally my intuition.

**Comment by Laurae2** [comment]
@goswami-rahul 

1. split data in train/test
2. train model on train, watch test scores
3. use the trained model to predict on test
4. create the confusion matrix of the prediction and the expected labels
5. report here the confusion matrix from step 4 and F1 score from step 3

To do with 1st iteration and last iteration. You can also compute manually the F1 score to compare using the confusion matrix.

**Comment by goswami-rahul** [comment]
@Laurae2 These are the results (f1 score is 0.69 in 1st iteration, 0.72 in 109 iteration)
```python
>>> lgbtrain = lgb.Dataset(train[200000:], label=target[200000:])
>>> lgbval = lgb.Dataset(train[:200000], target[:200000])
>>> booster = lgb.train(alg.get_params(), lgbtrain, num_boost_round=3000, evals_result=eval_res, valid_sets=lgbval,
                 early_stopping_rounds=80, verbose_eval=True, feval=f1_scorer)

>>> pred1 = booster.predict(train[:200000], num_iteration=1)
>>> pred109 = booster.predict(train[:200000], num_iteration=-1)
>>> print(metrics.confusion_matrix(target[:200000], pred1.argmax(axis=1)))
[[15322  3222   483   379    20]
 [  344 13977  6703  6020     8]
 [  141 10988 10908 16608    19]
 [  157  3791  8159 35909   152]
 [   32   101   266  2984 63307]]
>>> print(metrics.confusion_matrix(target[:200000], pred109.argmax(axis=1)))
[[15974  2429   956    67     0]
 [  308 12569  9820  4351     4]
 [   55  8132 15680 14782    15]
 [    7  2511  9337 36203   110]
 [    4    71   292  2927 63396]]
>>> print(metrics.f1_score(target[:200000], pred1.argmax(axis=1), average="weighted"))
0.6940951719647599
>>> print(metrics.f1_score(target[:200000], pred109.argmax(axis=1), average="weighted"))
0.7207151285072758
```

**Comment by Laurae2** [comment]
ping @StrikerRUS @wxchan - see previous and first post.

**Comment by StrikerRUS** [comment]
About the shape of `y_pred` in multiclass case:
https://github.com/Microsoft/LightGBM/blob/a6e878e2fc6e7f545921cbe337cc511fbd1f500d/python-package/lightgbm/sklearn.py#L80-L81

**Comment by fl2o** [comment]
Coming late but yes, you should reshape like this inside the feval for multiclass problems:
```
    labels = train_data.get_label()
    pred = pred.reshape(len(np.unique(labels)), -1)
```

**Comment by goswami-rahul** [comment]
@fl2o @StrikerRUS Thanks for the clarification.  
I think this information should be in the `feval` [docs](http://lightgbm.readthedocs.io/en/latest/Python-API.html#lightgbm.cv), could be helpful.
